### PR TITLE
[wasm] Purge Wasm when ECDS comes instead of triggering it by Timer

### PR DESCRIPTION
**Pull request by ingwonsong**
fixes: https://github.com/istio/istio/issues/42154

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@zirain was working on this

**Comment by ingwonsong** [comment]
@hzxuzhonghu yes. As mentioned in https://github.com/istio/istio/pull/42486#issuecomment-1363483556, let's try this PR first as a short term solution.

File: pkg/wasm/cache.go [hunk 1]
```diff
@@ -409,36 +408,26 @@ func (c *LocalFileCache) getEntry(key cacheKey, ignoreResourceVersion bool) (str
...
-				} else {
-					for downloadURL := range m.referencingURLs {
-						delete(c.checksums, downloadURL)
-					}
-					delete(c.modules, k)
-					wasmLog.Debugf("successfully removed stale Wasm module %v", m.modulePath)
-				}
+	c.mux.Lock()
+	for k, m := range c.modules {
+		if !m.expired(c.ModuleExpiry) {
```
Comment (zirain) [inline pkg/wasm/cache.go:415]: what will happen, if there's no xds update in 24h?

Comment (zirain) [reply pkg/wasm/cache.go:415]: IIRC, if ecds is cached in envoy, reconnect will not send ECDS request to xdsserver?

Comment (ingwonsong) [reply pkg/wasm/cache.go:415]: If no ECDS, no purgin will happen.

`purge` function is only called at the end of `Get` function.

Comment (ingwonsong) [reply pkg/wasm/cache.go:415]: > IIRC, if ecds is cached in envoy, reconnect will not send ECDS request to xdsserver?

Yes. ECDS request will be not sent to xdsserver if Envoy already has proper resources.

**Review by hzxuzhonghu — COMMENTED** [review]
my original thought was to infer the ecds resources from ecds requests, super sure this is not.

**Comment by ingwonsong** [comment]
I realized that this may not work for multiple extensions. Close this.

**Pull request closed without merging**